crates/bevy_app/src/terminal_ctrl_c_handler.rs

```diff
@@ -0,0 +1,63 @@
+use std::sync::atomic::{AtomicBool, Ordering};
+
+use bevy_ecs::event::EventWriter;
+
+use crate::{App, AppExit, Plugin, Update};
+
+#[cfg(not(target_arch = "wasm32"))]
+/// Indicates that all [`App`]'s should exit.
+static RECEIVED_CTRL_C: AtomicBool = AtomicBool::new(false);
+
+/// Gracefully handles `Ctrl+C` by emitting a [`AppExit`] event. This plugin is part of the `DefaultPlugins`.
+///
+/// ```no_run
+/// # use bevy_app::{App, NoopPluginGroup as MinimalPlugins, PluginGroup, TerminalCtrlCHandlerPlugin};
+/// fn main() {
+///     App::new()
+///         .add_plugins(MinimalPlugins)
+///         .add_plugins(TerminalCtrlCHandlerPlugin)
+///         .run();
+/// }
+/// ```
+///
+/// If you want to setup your own `Ctrl+C` handler, you should disable this
+/// plugin from `DefaultPlugins`:
+/// ```no_run
+/// # use bevy_app::{App, NoopPluginGroup as DefaultPlugins, PluginGroup, TerminalCtrlCHandlerPlugin};
+/// fn main() {
+///     App::new()
+///         .add_plugins(DefaultPlugins.build().disable::<TerminalCtrlCHandlerPlugin>())
+///         .run();
+/// }
+/// ```
+#[derive(Default)]
+pub struct TerminalCtrlCHandlerPlugin;
+
+#[cfg(not(target_arch = "wasm32"))]
+impl TerminalCtrlCHandlerPlugin {
+    /// Handler for when the user presses `Ctrl+C` on the terminal. Should be called from the `Ctrl+C` hook.
+    pub fn on_ctrl_c() {
+        RECEIVED_CTRL_C.store(true, Ordering::Relaxed);
+    }
+
+    /// Sends a [`AppExit`] event when the user presses `Ctrl+C` on the terminal.
+    fn exit_on_ctrl_c(mut events: EventWriter<AppExit>) {
+        if RECEIVED_CTRL_C.load(Ordering::Relaxed) {
+            events.send(AppExit::from_code(130));
+        }
+    }
+}
+
+impl Plugin for TerminalCtrlCHandlerPlugin {
+    fn build(&self, app: &mut App) {
+        #[cfg(not(target_arch = "wasm32"))]
+        {
+            ctrlc::set_handler(move || {
```

Comment:
could you use `try_set_handler` instead, and handle errors to log if there's already one set instead of panicking?

Reply:
`set_handler` always overwrites the current handler so it should not panic.
But I am changing it to `try_set_handle` so that the user can set their own handler. They can then call our `TerminalCtrlCHandlerPlugin::gracefully_exit` in their handler to exit bevy gracefully.